### PR TITLE
[NestedTensor] Integrate the softmax operator along the jagged dimension into NestedTensor

### PR DESCRIPTION
**Pull request by jananisriram**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131519
* __->__ #131518

Modify the existing `softmax` operator in PyTorch, invoked by `torch.softmax`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the aten padding operator, enables PyTorch users to invoke `torch.softmax` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` nested tensor.

Write unit tests based on the `sum` and `mean` jagged operators to verify the accuracy of the ragged reduction implementation for `torch.softmax`. Add unit tests to verify error handling for unsupported features in `NestedTensor` `torch.softmax`.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. In addition, the `softmax` operator is required to take in as input an integer for the reduction dimension `dim`, requiring new unit tests heavily inspired by the `sum` and `mean` jagged operator unit tests. `Softmax` also allows for reducing along the batch dimension.

### CODE REVIEW

**Comment by jananisriram** [comment]
@jananisriram has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59951075).

**Comment by jananisriram** [comment]
@jananisriram has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59951075).

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -230,7 +230,7 @@ def layout_name(layout):
 
 def get_op_name(layout):
     # e.g. "<OpOverload(op='aten.sum', overload='dim_IntList')>" -> "sum"
-    return layout.__name__.split(".")[0]
+    return layout.__name__.split(".")[0].split("_")[-1]
```
Comment (davidberard98) [inline test/test_nestedtensor.py:233]: curious why you need the `_` ?

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -4125,6 +4127,74 @@ def test_jagged_op_dim(
...
+        reduce_dims = (
+            (0, 0),
+            (2, 1),
+            (3, 2),
+        )  # (reduction dimension, effective reduction dimension for baseline)
+
+        for reduce_dim, _ in reduce_dims:
+            nt = torch.nested.as_nested_tensor(ts, layout=torch.jagged)
+            out_actual = torch.nn.functional.softmax(nt, dim=reduce_dim)
+            torch._dynamo.disable(self.assertEqual)(
```
Comment (davidberard98) [inline test/test_nestedtensor.py:4159]: nice solution!

File: torch/nested/_internal/ops.py [hunk 3]
```diff
@@ -502,16 +509,76 @@ def zero__default(func, *args, **kwargs):
...
+        new_kwargs["dim"],
+        reduce_on_batch,
+        reduce_on_ragged,
+        reduce_on_non_batch,
+    ) = _wrap_jagged_dims(
+        inp.dim(),
+        (new_kwargs["dim"],),
+        "softmax",
+        inp._ragged_idx,
+        allow_batch_dim=True,
```
Comment (davidberard98) [inline torch/nested/_internal/ops.py:540]: I don't think this is correct. We should probably error if we try to reduce on dim=0.

To check this, I think you can do something like this:
```
x = torch.randn(4, 4, 4)
nt = torch.nested.as_nested_tensor(x, layout=torch.jagged)
dense_softmax = torch.nn.functional.softmax(x, dim=0)
nt_softmax = torch.nn.functional.softmax(nt, dim=0)
assert torch.allclose(dense_softmax, nt_softmax.values())
```

**Review by davidberard98 — CHANGES_REQUESTED** [review]

File: test/test_nestedtensor.py (hunk 1 above)
Comment (jananisriram) [reply test/test_nestedtensor.py:233]: depending on the op, `layout.__name__.split(".")[0]` returns `_softmax` rather than `softmax`, so wanted to remove the `_`!

File: torch/nested/_internal/ops.py (hunk 3 above)
Comment (jananisriram) [reply torch/nested/_internal/ops.py:540]: noted, and added a test/assert for this!

File: torch/nested/_internal/ops.py [hunk 4]
```diff
@@ -36,14 +36,18 @@ def _wrap_jagged_dim(
     return _outer_to_inner_dim(ndim, wrapped) if convert_to_inner_dim else wrapped
 
 
-def _wrap_jagged_dims(ndim, dims, op_name, ragged_idx=1):
+def _wrap_jagged_dims(ndim, dims, op_name, ragged_idx=1, allow_batch_dim=False):
```
Comment (davidberard98) [inline torch/nested/_internal/ops.py:39]: nit: since you aren't using allow_batch_dim, you could probably remove the changes to this file

File: torch/nested/_internal/ops.py [hunk 5]
```diff
@@ -502,16 +509,80 @@ def zero__default(func, *args, **kwargs):
 
 
 @register_jagged_func(
-    torch.ops.aten._softmax.default, "self: jt, dim: any, half_to_float: any"
+    torch.ops.aten._softmax.default, "self: jt_all, dim: any?, half_to_float: any"
```
Comment (davidberard98) [inline torch/nested/_internal/ops.py:512]: nit: why is `dim` changed to `any?`? is dim optional?

**Review by davidberard98 — APPROVED** [review]
a few nits, otherwise LGTM! (also pending merge tests etc.)

File: torch/nested/_internal/ops.py (hunk 5 above)
Comment (jananisriram) [reply torch/nested/_internal/ops.py:512]: that's a good point; i was getting some errors with `dim` being non-optional for `sum` and `mean` so wanted to prevent that for `softmax`, but it should definitely require a `dim`!

File: torch/nested/_internal/ops.py (hunk 4 above)
Comment (jananisriram) [reply torch/nested/_internal/ops.py:39]: thanks! removed the condition for `allow_batch_dim`

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
Note: Please do not merge this PR.  Create a new codev diff and only use the codev tools to push the changes to github.  Alternatively, you can create the new PR directly on github and the pytorch diff train will take care of get it merged internally
